### DISCUSSION TITLE
Find the subject of a sentence given the verb and noun: POS_ tagging (avoiding noun-chunks!)

### DISCUSSION THREAD

**Discussion by wagnertail** [explosion/spaCy#9490] [Help: Coding & Implementations] [answered]
Example: I have a report, and from that report I want to pick out all of the names that have paid certain fines.  Furthermore, I would need to consider the fact that many sentences may be phrased differently:

'Michael paid a £10,000 fine.'
'Samantha paid a fine of £200.'
'John paid £2,000 in fines.'

I would like to find the best way to find the subject of the sentence (PERSON entity) given the MONEY entity and VERB of the sentence  

Task: 

1. Using NER, I would find all MONEY entities and all PERSON entities

```
money_entities = [ent for ent in sentence.ents if ent.label_ == "MONEY"]
person_entities = [ent for ent in sentence.ents if ent.label_ == "PERSON"]
```

2. For each of the MONEY entities, I would then find its root.head

```
for ent in money_entities:
    head = ent.root.head
    print(head)
```

3. If root.head of the MONEY entity is a VERB, I would like to know if there is a simple way to specify finding the 'nsubj' without having to crawl back word by word in the sentence

```
heads_list = [] # list of heads of MONEY entities
nsubj = [w for w in heads_list.lefts if w.dep_ == "nsubj"]
```

![image](https://user-images.githubusercontent.com/68034418/137633586-892aded4-8564-4d67-a219-936061406f47.png)

4. If root.head of the MONEY entity is a direct object 'dobj' (i.e. a NOUN), I would then have to find the head of ent.root.head to find the VERB.  

![image](https://user-images.githubusercontent.com/68034418/137633639-4314b0a0-bab9-4bcc-acb0-d0ab90eb3fd5.png)

5. If the root.head of the MONEY entity is a PREPOSITION, I would need to find the head of the head of ent.root.head 

![image](https://user-images.githubusercontent.com/68034418/137633684-3e2f8361-46d7-4d85-80b9-9f1ef1d9e474.png)

- ent.root.head: 'of' (PREPOSITION)
- ent.root.head.head: 'fine' ('dobj', NOUN)
- ent.root.head.head.head: 'paid' (VERB)

If there was a way a to be able to specify what part of the sentence I want to find given the verb, I would be able to instantly select the 'nsubj'.  Ideally, the task would be:

1) find all sentences with MONEY and PERSON entities
2) for each MONEY entity, find the related verb (via root.head and possibly further heads)
3) then once the verb is discovered, it would be great to have some sort of query-code to find 'nsubj' rather than having to go backwards word-by-word:

`nsubj = [w for w in head.lefts if w.dep_ == "nsubj"]`

4. then if the 'nsubj' is a PERSON entity (not a pronoun), I would want to check it against my list of PERSON entities, and pair with a MONEY entity then add to a DataFrame.  _I'm having problems with comparing the 'nsubj' variable with my list of PERSON entities as the former is a token whereas the list of PERSON entities is a list of span objects... Is there a solution without having to stringify the values?_

In an ideal world, there would be a query I could write to say:

'for each sentence, if MONEY entity and PERSON entity share VERB (or if both entities are 'connected' as subject and object, add entities to dataframe'...

Hope this makes sense! Thank you 🙏

**Comment by wagnertail** [comment]
If at all possible, I would like to do the following:

1. Avoid having to scan sentence by sentence. Instead, tokenise the complete report and then run NER.
2. Focus on clauses that include MONEY entities 
3. Then for each clause associated with a MONEY entity, I would identify if it is 'attached' to a PERSON entity
4. Then analysing the PERSON entity, I would check if it is the subject 

Further conditions might be that the verb is similarly to 'paid' (and other versions/ lemmatisation)

I would also have to consider passive phrases like:

'A fine of £100 was paid by Philip.'

**Comment by polm** [comment, marked as answer]
> In an ideal world, there would be a query I could write to say:

> 'for each sentence, if MONEY entity and PERSON entity share VERB (or if both entities are 'connected' as subject and object, add entities to dataframe'...

Uh, have you seen the [DependencyMatcher](https://spacy.io/usage/rule-based-matching#dependencymatcher)? It sounds like what you want is a rather straightforward application of it.

If you want to work with entities as single tokens you can use the [`merge_entities` pipe](https://spacy.io/api/pipeline-functions#merge_entities).

**Reply by wagnertail** [reply]
Superstar ⭐️ Thank you 🙏
